README.md
# QuinLED
This repository provides various bits of code, for use with the awesome [QuinLED PWM LED dimmer](http://blog.quindorian.org/2016/07/esp8266-lighting-revisit-and-history-of-quinled.html/), as created by Quindor. Please note that this is **not** the original code provided by Quindor, so if you got the code from this GitHub repository and have issues, please complain to me, not to Quindor. The original code has been altered and is provided here with permission from Quindor.

## Introduction
In short, the QuinLED module is a DIY Pulse-Width Modulation (PWM) LED dimmer. It is designed by Quindor, who has many [blog articles](http://blog.quindorian.org/2016/07/esp8266-lighting-revisit-and-history-of-quinled.html/) and [videos](https://www.youtube.com/playlist?list=PL4b74vD-Uo-NrEv0RkJdh8M4dz8zxxMBb) about the evolution of this dimmer. I highly recommend that you check these resources if you're unfamiliar with the QuinLED module.

<p align="center"><img src="misc/images/QuinLED.png" width="750"></p>

## Contents
While creating and using my own QuinLED modules, I noticed some small issues in the code provided. Since I'm familiar with Lua coding, I decided to fix these issues for myself, and improve the script while I was at it. I also added some extras, and will guide you through creating your own NodeMCU firmware.

I would suggest that you follow the links below in the order they're provided, this should make it easier to understand what's going on.

What is provided:
- [A guide on creating your own NodeMCU firmware for QuinLED](NodeMCU%20Firmware/README.md)  
  (This is completely optional, you can also use the firmware build provided by Quindor in the [Workshop Archive](http://blog.quindorian.org/2016/09/esp8266-led-lighting-setting-voltage-and-flashing-nodemcu.html/), or provided by [me](NodeMCU%20Firmware/nodemcu-master-6-modules-2018-05-18-18-09-02-integer.bin))
  - This way you can build your own firmware when you want to develop extra functionality for QuinLED, without guessing which modules you need.
- [Completely revamped program code](Program%20code/README.md)
  - Fixes reboot when providing a brightness level outside the 0-1023 range
  - Fixes reboot when providing a non-four-digit fadetime
  - Adds flexibility for future development with more channels
  - Adds a web server, which allows updating the code Over The Air (OTA), so no need to attach wires after the first time!
- [A different way of implementing QuinLED in Domoticz](Domoticz/README.md) (using dzVents)  
  (Again, completely optional. You can also use [Quindor's method](http://blog.quindorian.org/2017/02/esp8266-led-lighting-using-quinled-with-domoticz.html/))
  - Enables all script editing to be done from the browser, through Domoticz. No command line needed
  - Eliminates the need for a script for every device, one script can handle all your devices
- [A batch file](Batch%20file/README.md), that allows you to control your QuinLED modules easily through a command line interface  
  (Also optional, this is just for ease of use)
  - Allows for quick and easy input, without having to remember the exact syntax of the command

## Conclusion
If you followed all the readme's as linked above, you should now have:
- A QuinLED module running my custom firmware and program code
- The ability to change the program code OTA
- An easy way to add new modules through the dzVents scripts
- A simple batch file that allows you to control the QuinLED modules through the command line quickly and easily.

## Contributing
If you want to contribute to this repository in any way shape or form, please feel free to do so! Some suggestions:
- Translations of the README files
- Code improvements, changes, or additions
- Well.. anything really

To keep things clean, I do have the repo set to protected, meaning that I will have to approve any pull request manually. Some guidelines:
- Use proper indentation
- Try to use the same coding style as provided
- Try to provided descriptive comments for at least each code block you add

If you want to discuss your idea before you start working on it (which I'd recommend), feel free to open an issue on this repository.

Changelog.md
# Changelog
This file lists all 'major' changes to the repository, ignoring small syntax fixes etc.

## 2018-06-25
- Enabled background execution, allowing commands to be executed in much quicker succession.
- Added dimming-ramping. Allows for adding a 'sensitivity curve' to the QuinLED dimmers. More information in the [Domoticz README file](Domoticz/README.md#resolution-ramping).

## 2018-05-22
Initial release


Program code/init.lua
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ --
-- User changeable values
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ --

-- Enter the SSID (name) of the WiFi network for 'ssid', and enter the matching 
--	 password for 'pwd'
wifi.setmode(wifi.STATION)
wifi.sta.config(
	{
		ssid = "WIFI_SSID",
		pwd = "WIFI_PASSWORD",
	}
)

-- If you want a static IP for this QuinLED module, set the IP address, subnet
--	 mask and gateway here.
wifi.sta.setip({
      ip = "STATIC_IP",
      netmask = "SUBNET_MASK",
      gateway = "GATEWAY_IP",
  }
)

-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ --
-- End of user changeable values, start of program
--   Don't change anything below this line
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ --


-- Initiate QuinLED script
local QuinLED = require("QuinLED")
-- Initiate WebIDE script
local IDE = require("OnlineIDE")

function handleStuff()
	srv=net.createServer(net.TCP) 
	srv:listen(43333, function(conn)

		conn:on("receive", function(sck, payload)
        print("Init.lua :: RECEIVE")
			if (string.find(payload, "^Fadetimer")) then
                print("Init.lua :: QUINLED")
				QuinLED.receive(sck, payload)
			else
                print("Init.lua :: IDE")
				IDE.receive(sck, payload)
			end
		end)

		conn:on("sent", function(sck)
        print("Init.lua :: SENT")
		    IDE.sent(sck)
	  	end)
	end)
end

function connected()
    print("Connected to WiFi")
    local ip, nm, gw = wifi.sta.getip()
    print("\tIP: "..tostring(ip))
    print("\tSubnet Mask: "..tostring(nm))
    print("\tGateway: "..tostring(gw))
end

-- If an IP is assigned, print the çontents of the 'connected' function, and create the server.
--	 Otherwise, register a callback for when an IP is assigned, and then execute those actions.
if (wifi.sta.status() == wifi.STA_GOTIP) then
    connected()
    handleStuff()
else
    print("Connecting to WiFi...")
    wifi.eventmon.register(wifi.eventmon.STA_GOT_IP, function()
        wifi.eventmon.unregister(wifi.eventmon.STA_GOT_IP)
        connected()
        handleStuff()
    end)
end


Domoticz/global_data.lua
return {
	helpers = {
		
		QL_COMMAND = "echo Fadetimer=%d, LED%d_target=%d | nc -w %d %s %s &",
		
		QL_PORT = "43333",
		QL_RESOLUTION = 1023,
		QL_TIMEOUT = 1,
		
		-- Linking switch name to global_data identifier
		--	REQUIRED
		QL_NAME = {
		    ["QuinLED Dimmer1"] = "DIMMER1",
		    ["QuinLED Dimmer2"] = "DIMMER2",
	    },
		
		-- IP address (String)
		--	REQUIRED
		QL_IP = {
	        DIMMER1 = "192.168.1.1",
	        DIMMER2 = "192.168.1.1",
        },

        -- Output channel (Integer)
        --	REQUIRED
        QL_CHANNEL = {
            DIMMER1 = 1,
            DIMMER2 = 2,
        },

        -- Fadetime (Integer) [ms]
        -- 	Enables you to set a custom fadetime per device.
        --  When an entry is missing, or is set to nil, the script will use the default value of zero.
        QL_FADETIME = {
            DIMMER1 = 0,
	        DIMMER2 = 0,
        },

        -- Fadetime override (Integer) [ms]
        -- 	Enables you to (temporarilly) override the fadetime per device.
        --  When an entry is missing, or is set to nil, nothing will be overridden.
        QL_FADETIME_OVERRIDE = {
            DIMMER1 = nil,
	        DIMMER2 = nil,
        },

        -- Resolution override (Integer) [ms]
        --	Enables you to set a maximum dimming level per device.
        --  When an entry is missing, or is set to nil, the script will use the default resolution value.
        QL_RESOLUTION_OVERRIDE = {
            DIMMER1 = nil,
            DIMMER2 = nil,
        },

        -- Resolution ramping (Float) (x^y, where y is defined in the QL_RESOLUTION_RAMP table below)
        -- Accepted values: Anything greater than zero.
        -- Recommended values: Anything between zero and five.
        
        -- For a visualisation of the dimming curve, go to https://www.desmos.com/calculator (or any other graphical calculator),
        -- and enter 'x^y' as the formula, where y is the number you enter in this table.
        -- When there is no entry for a device, x^1 is used, which results in a straight line.
        
        -- (0 < y < 1) results in a more 'agressive' dimming (gets brighter earlier, fine-tuning in the end)
        -- (1) results in a linear dimming.
        -- (1 < y < inf.) results in a more 'relaxed' dimming (fine-tuning in the beginning, gets brighter quickly at the end)
        QL_RESOLUTION_RAMP = {
            DIMMER1 = 1.2,
            DIMMER2 = 2,
        },
	}
}


Program code/QuinLED.lua
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ --
-- User changeable values
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ --

-- Add new channels in PWM_id. The key relates to the number in 'LED1_target' in the command, the 
--	 value relates to the pin number on the hardware
local PWM_id = {
    [1]=3, 
    [2]=4,
}

-- Levels at which the PWM dimming starts when QuinLED boots. [0-1023]
local startupLevels = {
    [1]=0, 
    [2]=0,
}

-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ --
-- End of user changeable values, start of program
--   Don't change anything below this line
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ --

-- For external access and global storage
local QuinLED = {}
local LED_current = {}
local LED_target = {}
local LED_timer = {}

-- For every entry in the PWM_id table, setup and start the PWM signal, and set the current and 
--	 and target level to zero.
for key,id in pairs(PWM_id) do
    local startupLevel = startupLevels[key]
    if (startupLevel == nil) then
        startupLevel = 0
    end
	pwm.setup(id, 1000, startupLevel)
	pwm.start(id)

	LED_current[key] = startupLevel
	-- 	LED_target[key] = 500
end

function QuinLED_Receive(conn, payload)
	-- This needs to be expanded when more channels are added, and should probably be improved
	--	but at the moment, I can't think of a better way to do this.
	local function startTimer(idx, dimInterval)

		-- Force stop the previous timer if it's active and unregister the callback (frees up resources)
		local timer = LED_timer[idx]
		timer:unregister()

		local function createCallback(idx)

			return (function()
					if (LED_current[idx] < LED_target[idx]) then 
						LED_current[idx] = (LED_current[idx] + 1) 
						pwm.setduty(PWM_id[idx], LED_current[idx])
					elseif (LED_current[idx] > LED_target[idx]) then 
						LED_current[idx] = (LED_current[idx] - 1) 
						pwm.setduty(PWM_id[idx], LED_current[idx])
					elseif (LED_current[idx] == LED_target[idx]) then 
						local timer = LED_timer[idx]
						timer:unregister()
					end 
				end)
		end

		local callback = createCallback(idx)
		
		local timer = tmr.create()
		timer:alarm(dimInterval, 1, callback)
		LED_timer[idx] = timer
	end
	print("\n > Payload: "..payload)

	-- Futureproofing
	--	Search for the index of the used LED in the payload, set to first match (max idx is 12)
	local LEDidx
	for i=1,12 do
		if (string.find(payload, string.format("LED%d", i)) ~= nil) then
			LEDidx = i
			break
		end
	end
 
	local LED = string.format("LED%d", LEDidx)
	local fadeTimeStartIdx,fadeTimeEndIdx = string.find(payload, "Fadetimer=%d+")
	local targetStartIdx,targetEndIdx = string.find(payload, LED.."_target=%d+")

	if (fadeTimeStartIdx ~= nil and fadeTimeEndIdx ~= nil and targetStartIdx ~= nil and targetEndIdx ~= nil) then

		-- Add the length of the strings 'Fadetimer=' and 'LED#_target=' to the start idx to get the number idx
		fadeTimeStartIdx = fadeTimeStartIdx + 10
		targetStartIdx = targetStartIdx + 12

		local target = string.sub(payload, targetStartIdx, targetEndIdx)
		print("RAW: LED_target: "..tostring(target))
		target = tonumber(target)
		if (target == nil) then
			print("Target is nil, possibly trying to convert text to number. Using default (5)")
			target = 5
		end
		-- Clamp values between 0 and 1023
		LED_target[LEDidx] = math.max(0, math.min(1023, target))

		local fadeTime = string.sub(payload, fadeTimeStartIdx, fadeTimeEndIdx)
		print("RAW: Fadetime: "..tostring(fadeTime))
		fadeTime = tonumber(fadeTime)
		if (fadeTime == nil) then
			print("Fadetime is nil, possibly trying to convert text to number. Using default (5000ms)")
			fadeTime = 5000
		end

		-- Number of steps between target and current
		local numSteps = math.abs(LED_target[LEDidx] - LED_current[LEDidx])

		-- If 0 steps, set to 1 to make sure division below doesn't crap out
		if (numSteps == 0) then
			numSteps = 1
		end

		local dimInterval = (fadeTime / numSteps)

		print ("Fadetime: "..fadeTime)
		print ("NumSteps: "..numSteps)
		print ("DimInterval: "..dimInterval)
		print ("Current LED value: "..LED_current[LEDidx])
		print ("Target LED value: "..LED_target[LEDidx])

		-- If Dimtime is less then 1ms, do instantly, otherwise set timer
		if (dimInterval < 1) then
			LED_current[LEDidx] = LED_target[LEDidx]
			pwm.setduty(PWM_id[LEDidx], LED_current[LEDidx])
		else
			startTimer(LEDidx, dimInterval)
		end
	end
end
QuinLED.receive = QuinLED_Receive

print ("QuinLED script loaded")

return QuinLED


Domoticz/QuinLED Dimmer.lua
return {

	active = {
		true
	},
	
	on = {
		devices = {
			'QuinLED Dimmer1',
			'QuinLED Dimmer2',
		},
	},
	
	logging = {
        -- level = domoticz.LOG_DEBUG,
        level = domoticz.LOG_ERROR,
        marker = "QL_Dimmer :",
    },

	execute = function(domoticz, device)
	    
	    -- "Global" stuff
	    local helpers = domoticz.helpers
		local command = helpers.QL_COMMAND
		
		-- Find the correct name
	    local deviceName = helpers.QL_NAME[device.name]
	    if (deviceName == nil or deviceName == "") then
	        domoticz.log("Devicename \""..device.name.."\" unknown.", domoticz.LOG_ERROR)
	        return
        end
	    
	    -- Handle brightness
		local targetLevel = 0
		if (device.active and device.level > 0) then
		    
		    -- Check if the resolution is overridden
		    local resolution = helpers.QL_RESOLUTION_OVERRIDE[deviceName]
		    if (resolution == nil) then
		        resolution = helpers.QL_RESOLUTION
	        end
	        
		    -- Get the normalized level [0-1]
		    --  Divide by 96 because the dummy dimmer always goes back to 96% for some reason
            local normTargetLevel = (device.level / 96)
		    
		    -- Apply the ramp setting, if available
		    local exponent = helpers.QL_RESOLUTION_RAMP[deviceName]
		    local rampedTargetLevel
		    if (exponent ~= nil and type(exponent) == "number") then
		        rampedTargetLevel = normTargetLevel^exponent
		        domoticz.log("Ramped target level: "..normTargetLevel.." ^ "..exponent.." = "..rampedTargetLevel)
	        end
	        
	        -- Set targetLevel to either the ramped or the normal level
	        if (rampedTargetLevel ~= nil) then
	            targetLevel = rampedTargetLevel * resolution
            else
                targetLevel = normTargetLevel * resolution
            end
            
            -- Add 4 (minimum value for the LED strip to light up) to make sure the
            --  light always lights up at the lowest setting above zero.
            targetLevel = targetLevel + 4
        end
        
        -- Handle fadetime
        local fadeTime = helpers.QL_FADETIME[deviceName]
        if (fadeTime == nil) then
            fadeTime = 0
        end
        
        -- Check if fadetime is overridden
        if (helpers.QL_FADETIME_OVERRIDE[deviceName] ~= nil) then
            fadeTime = helpers.QL_FADETIME_OVERRIDE[deviceName]
        end
        
        -- Build command
        command = string.format(
            command,
            fadeTime,
            helpers.QL_CHANNEL[deviceName],
            targetLevel,
            helpers.QL_TIMEOUT,
            helpers.QL_IP[deviceName],
            helpers.QL_PORT
        )
        
        -- Execute command
        os.execute(command)
        domoticz.log("Command: "..tostring(command))
	end
}



Domoticz/QuinLED Switch.lua
return {

	active = {
		true
	},
	
	on = {
		devices = {
            'QuinLED Switch1',
            'QuinLED Switch2',
		},
	},
	
	logging = {
        -- level = domoticz.LOG_DEBUG,
        marker = "QL_Fan ::",
    },

	execute = function(domoticz, device)
	    
		local command = domoticz.helpers.QL_COMMAND
		
		-- Start at a targetLevel of zero, set to maximum if the switch was turned on
		local targetLevel = 0
		if (device.active) then
            targetLevel = domoticz.helpers.QL_RESOLUTION
        end
        
        -- Check if fadetime is overridden
        local fadeTime = domoticz.helpers.QL_FAN_FADETIME_DEFAULT
        if (domoticz.helpers.QL_FAN_FADETIME_OVERRIDE ~= nil) then
            fadeTime = domoticz.helpers.QL_FAN_FADETIME_OVERRIDE
        end
        
        -- Build command
        command = string.format(
            command,
            fadeTime,
            domoticz.helpers.QL_FAN_CHANNEL,
            targetLevel,
            domoticz.helpers.QL_FAN_WAITTIME,
            domoticz.helpers.QL_FAN_IP,
            domoticz.helpers.QL_PORT
        )
        
        -- Execute command
        os.execute(command)
	end
}

Program code/OnlineIDE.lua
local IDE = {}

IDE.rnrn = 0
IDE.Status = 0
IDE.DataToGet = 0
IDE.method = ""
IDE.url = ""
IDE.vars = ""

function EditorReceive(sck, payload, aceEnabled)

	local AceEnabled = aceEnabled == nil and true or aceEnabled

	if IDE.Status == 0 then
        _, _, IDE.method, IDE.url, IDE.vars = string.find(payload, "([A-Z]+) /([^?]*)%??(.*) HTTP")
        -- print("IDE.method, IDE.url, IDE.vars: ", IDE.method, IDE.url, IDE.vars)
    end
    
    if IDE.method == "POST" then
    
        if IDE.Status == 0 then
            -- print("IDE.Status", IDE.Status)
            _, _, IDE.DataToGet, payload = string.find(payload, "Content%-Length: (%d+)(.+)")
            if IDE.DataToGet then
                IDE.DataToGet = tonumber(IDE.DataToGet)
                -- print("IDE.DataToGet = "..IDE.DataToGet)
                IDE.rnrn = 1
                IDE.Status = 1                
            else
                print("bad length")
            end
        end
        
        -- find /r/n/r/n
        if IDE.Status == 1 then
            -- print("IDE.Status", IDE.Status)
            local payloadlen = string.len(payload)
            local mark = "\r\n\r\n"
            local i
            for i=1, payloadlen do                
                if string.byte(mark, IDE.rnrn) == string.byte(payload, i) then
                    IDE.rnrn = IDE.rnrn + 1
                    if IDE.rnrn == 5 then
                        payload = string.sub(payload, i+1, payloadlen)
                        file.open(IDE.url, "w")
                        file.close() 
                        IDE.Status = 2
                        break
                    end
                else
                    IDE.rnrn = 1
                end
            end    
            if IDE.Status == 1 then
                return 
            end
        end       
    
        if IDE.Status == 2 then
            -- print("IDE.Status", IDE.Status)
            if payload then
                IDE.DataToGet = IDE.DataToGet - string.len(payload)
                --print("IDE.DataToGet:", IDE.DataToGet, "payload len:", string.len(payload))
                file.open(IDE.url, "a+")
                file.write(payload)            
                file.close() 
            else
                sck:send("HTTP/1.1 200 OK\r\n\r\nERROR")
                IDE.Status = 0
            end

            if IDE.DataToGet == 0 then
                sck:send("HTTP/1.1 200 OK\r\n\r\nOK")
                IDE.Status = 0
            end
        end
        
        return
    end
    -- end of POST IDE.method handling
    
    IDE.DataToGet = -1
    
    if IDE.url == "favicon.ico" then
        -- print("favicon.ico handler sends 404")
        sck:send("HTTP/1.1 404 file not found")
        return
    end    

    local sen = "HTTP/1.1 200 OK\r\n\r\n"
    
    -- it wants a file in particular
    if IDE.url ~= "" and IDE.vars == "" then
        IDE.DataToGet = 0
        sck:send(sen)
        return
    end

    sen = sen .. "<html><body><h1><a href='/'>NodeMCU IDE</a></h1>"
    
    if IDE.vars == "edit" then
        if AceEnabled then
            local mode = 'ace/mode/'
            if IDE.url:match(".css") then mode = mode .. 'css'
            elseif IDE.url:match(".html") then mode = mode .. 'html'
            elseif IDE.url:match(".json") then mode = mode .. 'json'
            elseif IDE.url:match(".js") then mode = mode .. 'javascript'
            else mode = mode .. 'lua'
            end
            sen = sen .. "<style type='text/css'>#editor{width: 100%; height: 80%}</style><div id='editor'></div><script src='//rawgit.com/ajaxorg/ace-builds/master/src-min-noconflict/ace.js'></script>"
                .. "<script>var e=ace.edit('editor');e.setTheme('ace/theme/monokai');e.getSession().setMode('"..mode.."');function getSource(){return e.getValue();};function setSource(s){e.setValue(s);}</script>"
        else
            sen = sen .. "<textarea name=t cols=79 rows=17></textarea></br>"
                .. "<script>function getSource() {return document.getElementsByName('t')[0].value;};function setSource(s) {document.getElementsByName('t')[0].value = s;};</script>"
        end
        sen = sen .. "<script>function tag(c){document.getElementsByTagName('w')[0].innerHTML=c};var x=new XMLHttpRequest();x.onreadystatechange=function(){if(x.readyState==4) setSource(x.responseText);};"
        .. "x.open('GET',location.pathname);x.send()</script><button onclick=\"tag('Saving, wait!');x.open('POST',location.pathname);x.onreadystatechange=function(){console.log(x.readyState);"
        .. "if(x.readyState==4) tag(x.responseText);};x.send(new Blob([getSource()],{type:'text/plain'}));\">Save</button> <a href='?run'>run</a> <w></w>"

    elseif IDE.vars == "run" then
        sen = sen .. "<verbatim>"

        function s_output(str) sen = sen .. str end
        node.output(s_output, 0) -- re-direct output to function s_output.

        local st, result = pcall(dofile, IDE.url)

        -- delay the output capture by 1000 milliseconds to give some time to the user routine in pcall()
        tmr.alarm(0, 1000, tmr.ALARM_SINGLE, function() 
            node.output(nil)
            if result then
                local outp = tostring(result):sub(1,1300) -- to fit in one send() packet
                result = nil
                sen = sen .. "<br>Result of the run: " .. outp .. "<br>"
            end
            sen = sen .. "</verbatim></body></html>"
            sck:send(sen)
        end)

        return

    elseif IDE.vars == "compile" then
        collectgarbage()
        node.compile(IDE.url)
        IDE.url = ""

    elseif IDE.vars == "delete" then
        file.remove(IDE.url)
        IDE.url = ""

    elseif IDE.vars == "restart" then
        node.restart()
        return

    end

    local message = {}
    message[#message + 1] = sen
    sen = nil
    if IDE.url == "" then
        local l = file.list();
        message[#message + 1] = "<table border=1 cellpadding=3><tr><th>Name</th><th>Size</th><th>Edit</th><th>Compile</th><th>Delete</th></tr>"
        for k,v in pairs(l) do
            local line = "<tr><td><a href='" ..k.. "'>" ..k.. "</a></td><td>" ..v.. "</td><td>"
            local editable = k:sub(-4, -1) == ".lua" or k:sub(-4, -1) == ".css" or k:sub(-5, -1) == ".html" or k:sub(-5, -1) == ".json"
            if editable then
                line = line .. "<a href='" ..k.. "?edit'>edit</a>"
            end
            line = line .. "</td><td>"
            if k:sub(-4, -1) == ".lua" then
                line = line .. "<a href='" ..k.. "?compile'>compile</a>"
            end
            line = line .. "</td><td><a href='" ..k.. "?delete'>delete</a></td></tr>"
            message[#message + 1] = line
        end
        message[#message + 1] = "</table><a href='#' onclick='v=prompt(\"Filename\");if (v!=null) { this.href=\"/\"+v+\"?edit\"; return true;} else return false;'>Create new</a> &nbsp; &nbsp; "
    message[#message + 1] = "<a href='#' onclick='var x=new XMLHttpRequest();x.open(\"GET\",\"/?restart\");x.send();setTimeout(function(){location.href=\"/\"},5000);this.innerText=\"Please wait\";return false'>Restart</a>"
    end
    message[#message + 1] = "</body></html>"

    local function send_table(sk)
        if #message > 0 then
            sk:send(table.remove(message, 1))
        else
            sk:close()
            message = nil
        end
    end
    sck:on("sent", send_table)
    send_table(sck)
end
IDE.receive = EditorReceive

function EditorSent(sck)
    if (IDE.DataToGet >= 0 and IDE.method == "GET") then
        if (file.open(IDE.url, "r")) then
            file.seek("set", IDE.DataToGet)
            local chunkSize = 512
            local line = file.read(chunkSize)
            file.close()
            if (line) then
                sck:send(line)
                IDE.DataToGet = IDE.DataToGet + chunkSize
                if (string.len(line) == chunkSize) then 
                	return
                end
            end
        end        
    end

	sck:close()
	sck = nil
end
IDE.sent = EditorSent

print ("OnlineIDE script loaded")

return IDE
